editor-integration/emacs/dune-mode.el

```diff
@@ -0,0 +1,328 @@
+;; WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
+;; AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
+;; CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
+;; LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
+;; NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+;; CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+(require 'scheme)
+
+(defvar dune-mode-hook nil
+  "Hooks for the `dune-mode'.")
+
+(defvar dune-program
+  (expand-file-name "dune-lint" dune-temporary-file-directory)
+  "Script to use to check the dune file.")
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;;;                     Syntax highlighting
+
+(defface dune-error-face
+  '((t (:foreground "yellow" :background "red" :bold t)))
+  "Face for errors (e.g. obsolete constructs).")
+
+(defvar dune-error-face 'dune-error-face
+  "Face for errors (e.g. obsolete constructs).")
+
+(defconst dune-keywords-regex
+  (eval-when-compile
+    (concat (regexp-opt
+             '("jbuild_version" "library" "executable" "executables" "rule"
+               "ocamllex" "ocamlyacc" "menhir" "alias" "install"
+               "copy_files" "copy_files#" "include")
+             ) "\\(?:\\_>\\|[[:space:]]\\)"))
+  "Keywords in dune files.")
+
+(defconst dune-fields-regex
+  (eval-when-compile
+    (regexp-opt
+     '("name" "public_name" "synopsis" "modules" "libraries" "wrapped"
+       "inline_tests" "inline_tests.backend"
+       "preprocess" "preprocessor_deps" "optional" "c_names" "cxx_names"
+       "install_c_headers" "modes" "no_dynlink" "kind"
+       "ppx_runtime_libraries" "virtual_deps" "js_of_ocaml" "flags"
+       "ocamlc_flags" "ocamlopt_flags" "library_flags" "c_flags"
+       "cxx_flags" "c_library_flags" "self_build_stubs_archive"
+       "modules_without_implementation"
+       ;; + for "executable" and "executables":
+       "package" "link_flags" "modes" "names" "public_names"
+       ;; + for "rule":
+       "targets" "action" "deps" "mode"
+       ;; + for "menhir":
+       "merge_into"
+       ;; + for "install"
+       "section" "files" "lib" "libexec" "bin" "sbin" "toplevel" "share"
+       "share_root" "etc" "doc" "stublibs" "man" "misc")
+     'symbols))
+  "Field names allowed in dune files.")
+
+(defvar dune-builtin-regex
+  (eval-when-compile
+    (concat (regexp-opt
+             '(;; Actions
+               "run" "chdir" "setenv"
+               "with-stdout-to" "with-stderr-to" "with-outputs-to"
+               "ignore-stdout" "ignore-stderr" "ignore-outputs"
+               "progn" "echo" "write-file" "cat" "copy" "copy#" "system"
+               "bash" "diff" "diff?"
+               ;; inline_tests and inline_tests.backend
+               ;; FIXME: "flags" is already a field and we do not have enough
+               ;; context to distinguishing both.
+               "backend" "generate_runner" "runner_libraries" "flags"
+               "extends")
```

Comment:
This is not up-to-date with the current Dune.  I'll do a PR with my WIP once this is merged but some help is welcome.